server/src/main/java/org/elasticsearch/search/aggregations/support/CoreValuesSourceType.java

```diff
@@ -224,7 +233,110 @@ public ValuesSource getField(FieldContext fieldContext, AggregationScript.LeafFa
+        }
+
+        @Override
+        public DocValueFormat getFormatter(String format, ZoneId tz) {
+            return DocValueFormat.IP;
+        }
+    },
+    DATE(EquivalenceType.NUMBER) {
+        @Override
+        public ValuesSource getEmpty() {
+            return NUMERIC.getEmpty();
+        }
+
+        @Override
+        public ValuesSource getScript(AggregationScript.LeafFactory script, ValueType scriptValueType) {
+            return NUMERIC.getScript(script, scriptValueType);
+        }
+
+        @Override
+        public ValuesSource getField(FieldContext fieldContext, AggregationScript.LeafFactory script) {
+            return NUMERIC.getField(fieldContext, script);
+        }
+
+        @Override
+        public ValuesSource replaceMissing(ValuesSource valuesSource, Object rawMissing, DocValueFormat docValueFormat, LongSupplier now) {
+            return NUMERIC.replaceMissing(valuesSource, rawMissing, docValueFormat, now);
+        }
+
+        @Override
+        public DocValueFormat getFormatter(String format, ZoneId tz) {
+            return  new DocValueFormat.DateTime(
+                format == null ? DateFieldMapper.DEFAULT_DATE_TIME_FORMATTER : DateFormatter.forPattern(format),
+                tz == null ? ZoneOffset.UTC : tz,
+                // If we were just looking at fields, we could read the resolution from the field settings, but we need to deal with script
+                // output, which has no way to indicate the resolution, so we need to default to something.  Milliseconds is the standard.
+                DateFieldMapper.Resolution.MILLISECONDS);
+        }
+    },
+    BOOLEAN(EquivalenceType.NUMBER) {
+        @Override
+        public ValuesSource getEmpty() {
+            return NUMERIC.getEmpty();
+        }
+
+        @Override
+        public ValuesSource getScript(AggregationScript.LeafFactory script, ValueType scriptValueType) {
+            return NUMERIC.getScript(script, scriptValueType);
+        }
+
+        @Override
+        public ValuesSource getField(FieldContext fieldContext, AggregationScript.LeafFactory script) {
+            return NUMERIC.getField(fieldContext, script);
+        }
+
+        @Override
+        public ValuesSource replaceMissing(ValuesSource valuesSource, Object rawMissing, DocValueFormat docValueFormat, LongSupplier now) {
+            return NUMERIC.replaceMissing(valuesSource, rawMissing, docValueFormat, now);
+        }
+
+        @Override
+        public DocValueFormat getFormatter(String format, ZoneId tz) {
+            return DocValueFormat.BOOLEAN;
+        }
+    }
+    ;
+
+    enum EquivalenceType {
+        STRING, NUMBER, GEO, RANGE, HISTOGRAM;
+    }
+
+    EquivalenceType equivalenceType;
+
+    CoreValuesSourceType(EquivalenceType equivalenceType) {
+        this.equivalenceType = equivalenceType;
+    }
+    @Override
+    public boolean isCastableTo(ValuesSourceType valuesSourceType) {
```

Comment:
This function,  along with the whole `EquivalenceType` enum exists to mimic the old `ValueType.isA` logic (and I re-implemented `isA` in terms of this function).  The whole concept right now is pretty dodgy, since it lets you treat booleans as dates for example, but removing it entirely in this PR is too much, so I settled for making it explicit here.  Definitely needs a follow up discussion, and possibly removal.